synnax/pkg/distribution/channel/channel_group.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2024 Synnax Labs, Inc.
+//
+// Use of this software is governed by the Business Source License included in the file
+// licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with the Business Source
+// License, use of this software will be governed by the Apache License, Version 2.0,
+// included in the file licenses/APL.txt.
+//
+// channelGroup implements tools for constructing a set of channels. The set of
+// channels is constructed as a sorted array of channelRanges. Each channelRange
+// consists of the local key that the range starts at and the size of the
+// channelRange (how many consecutive keys it is).
+
+package channel
+
+import (
+	"slices"
+	"sort"
+)
+
+// channelRange is the basic struct for ChannelGroup, and is defined as a set of
+// consecutive channel keys.
+type channelRange struct {
+	// start is the local key that the channelRange starts at.
+	start uint16
+	// size is how many consecutive keys there are in a row.
+	size uint16
+}
+
+// end returns the key that is at the end of channelRange r
+func (r channelRange) end() uint16 {
+	return r.start + r.size - 1
+}
+
+// ChannelGroup extends []channelRange with a few convenience methods.
+type ChannelGroup []channelRange

```

Comment:
This should be called `Group` and not `ChannelGroup`. Also, I feel like group is an ambiguous name. Isn't this more of a set?

Reply:
Also does this struct need to be public to the package? Shouldn't it be internal to the channel package.

Reply:
If you want to be able to test this, move it to an internal package inside the `channel` package